dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ExceptionDeserializingErrorDecoder.java

```diff
@@ -0,0 +1,299 @@
+        }
+        try {
+            return decodeInternal(response);
+        } catch (Exception e) {
+            e.addSuppressed(diagnostic(response));
+            throw e;
+        }
+    }
+
+    private @Nullable String extractErrorName(byte[] body) {
+        try {
+            NamedError namedError = NAMED_ERROR_DESERIALIZER.deserialize(new ByteArrayInputStream(body));
+            if (namedError == null) {
+                return null;
+            }
+            return namedError.errorName();
+        } catch (IOException | RuntimeException e) {
+            return null;
+        }
+    }
+
+    Optional<RuntimeException> checkCode(Response response) {
+        int code = response.code();
+        switch (code) {
+            case 308 -> {
+                Optional<String> location = response.getFirstHeader(HttpHeaders.LOCATION);
+                if (location.isPresent()) {
+                    String locationHeader = location.get();
+                    try {
+                        UnknownRemoteException remoteException = new UnknownRemoteException(code, "");
+                        remoteException.initCause(
+                                QosException.retryOther(qosReason(response), new URL(locationHeader)));
+                        return Optional.of(remoteException);
+                    } catch (MalformedURLException e) {
+                        log.error(
+                                "Failed to parse location header for QosException.RetryOther",
+                                UnsafeArg.of("locationHeader", locationHeader),
+                                e);
+                    }
+                } else {
+                    log.error("Retrieved HTTP status code 308 without Location header, cannot perform "
+                            + "redirect. This appears to be a server-side protocol violation.");
+                }
+            }
+            case 429 -> {
+                return Optional.of(response.getFirstHeader(HttpHeaders.RETRY_AFTER)
+                        .map(Longs::tryParse)
+                        .map(Duration::ofSeconds)
+                        .map(duration -> QosException.throttle(qosReason(response), duration))
+                        .orElseGet(() -> QosException.throttle(qosReason(response))));
+            }
+            case 503 -> {
+                return Optional.of(QosException.unavailable(qosReason(response)));
+            }
+        }
+        return Optional.empty();
+    }
+
+    private RuntimeException decodeInternal(Response response) {
+        // Check if the response code indicates an error, and if so, handle it accordingly.
+        Optional<RuntimeException> maybeException = checkCode(response);
+        if (maybeException.isPresent()) {
+            return maybeException.get();
+        }
+        int code = response.code();
+        byte[] body;
+        try {
+            body = toByteArray(response.body());
```

Comment:
style nit:
```suggestion
        byte[] body;
        try (InputStream bodyStream = response.body()) {
            body = bodyStream.readAllBytes();
```
avoids an extra method and doesn't really change readability